Fix DDMC->IMC block-block transition, for coarse(r) current block; fix DDMC near-hang pathology.

RyanWollaeger opened the pull request:
## Background

* A DDMC block transition issue was caught during ongoing testing with the large shearing box problem.
* Also, mitigate apparent near-hangs with DDMC resulting from misapplication of `ptcl_ddmc_albedo`.

## Description of Changes

+ Check for DDMC block-transition 0-velocity indicator in IMC cell.
+ Use `ptcl_ddmc_to_imc` function to resample position and direction.
+ Modify if-checks in `ptcl_ddmc_albedo` to avoid application to particles with invalid direction.
+ Restrict volume source particle sampling to a tolerance inside cell that avoids false application of `ptcl_ddmc_albedo`
+ Force particles to slightly past end of time step when within a tolerance of the end (IMC) or if census is sampled (DDMC).
+ (Also) update ~and activate~ swarm defrag task.

## Checklist

- [ ] New features are documented
- [ ] Tests added for bug fixes and new features
- [ ] (@lanl.gov employees) Update copyright on changed files

RyanWollaeger committed:
Fix DDMC->IMC block-block transition, for coarse(r) current block.

+ Check for DDMC block-transition 0-velocity indicator in IMC cell.
+ Use ptcl_ddmc_to_imc function to resample position and direction.
+ (Also) update and activate swarm defrag task.

Note: this commit is an incomplete draft.

pdmullen commented on src/jaybenne/jaybenne.cpp:
```diff
@@ -164,6 +164,9 @@ TaskCollection RadiationStep(Mesh *pmesh, const Real t_start, const Real dt) {
     auto control_pop = tl.AddTask(update_fluid, jaybenne::ControlPopulation, base.get());
+
+    // Defrag particles?
+    auto defrag_pop = tl.AddTask(control_pop, jaybenne::DefragParticles, base.get());
```
Thanks for adding this in! I have not experimented with defragging yet.  Would be good to ensure it is working as expected (maybe you have already done this?)

RyanWollaeger replied:
Not too much yet - I tried a couple of values for the threshold parameter, but did not notice a significant difference in run time per cycle (excluding the sporadic near-hanging cycles).

pdmullen replied:
So the observation is that it doesn't improve performance a whole lot, except for in near-hanging cycles?

I am a little hesitant to recommend defragging for calculations at this point.  I am not sure I trust it quite yet (it is also doing stuff under the hood manipulating `SwarmPack` caches).

Can we make it default to off? And maybe we can hand off the utility to Alex to (1) make sure it is behaving correctly and (2) figure out how to utilize it in the most effective way possible for performance?

alexrlongne replied:
I'm happy to take a look at it!

RyanWollaeger replied:
No - this does not help with near-hangs - sorry for the confusion here. Activated it to test if it had an effect on near-hangs (initially it seemed to, but the run I think merely got lucky). Left it in here mainly to point to it as a discussion topic - happy to deactivate it here.

RyanWollaeger replied:
Commented this line out and added a TODO here, to check SwarmPack defrag mechanics.

pdmullen commented on src/jaybenne/jaybenne.cpp:
```diff
@@ -662,13 +665,26 @@ TaskStatus UpdateDerivedTransportFields(MeshData<Real> *md, const Real dt) {
+  auto vmesh = desc.GetPack(md);
+
+  const int &nblocks = vmesh.GetNBlocks();
   auto &min_swarm_occupancy = jbn->template Param<Real>("min_swarm_occupancy");
```
```suggestion
TaskStatus DefragParticles(MeshData<Real> *md) {
  auto pm = md->GetParentPointer();
  auto &jbn = pm->packages.Get("jaybenne");
  auto &min_swarm_occupancy = jbn->template Param<Real>("min_swarm_occupancy");
  const int nblocks = md->NumBlocks();
  
```

pdmullen commented on src/jaybenne/transport_utils.hpp:
```diff
@@ -141,6 +141,13 @@ void ptcl_transport_step(tran_step_args tra) {
   const Real dt_push =
       ((tra.is_absorbed) ? dx_abs : ((tra.is_scattered) ? dx_sc : dx_push)) / tra.vv;
 
+  if (!(dt_push > 0.0) || !(dx_push > 0.0)) {
```
Need `printf` on device... but I think that here we should actually do something like `PARTHENON_WARN` instead.

RyanWollaeger replied:
These were only intended for testing - but that is a good idea to maybe make some warning checks.

pdmullen replied:
:+1:

I am also okay with just removing them if you like.

RyanWollaeger replied:
Removed this

pdmullen commented on src/jaybenne/transport_ddmc.cpp:
```diff
@@ -194,6 +195,12 @@ TaskStatus TransportPhotons_DDMC(MeshData<Real> *md, const Real t_start, const R
                                   t, x, y, z, is_absorbed, is_scattered};
+
+              // if velocity is 0, particle is from a DDMC cell in another block at <= refinement
+              if (vx * vx + vy * vy + vz * vz < 0.5 * vv * vv) ptcl_ddmc_to_imc(tra);
```
Can you help me understand the crux of the issue here? (sorry it's not immediately obvious).  This is just an interesting code path... wondering if there is another indicator other than the |v|?

RyanWollaeger replied:
We should maybe have an extended discussion about this, but the main idea is that when the particle speed is 0, it means it was a DDMC particle on another block in the previous iteration. However, this is only true of the previous block was at the same or higher resolution; otherwise the `sample_ddmc_bface` task would have sampled a refined face into the current block and reset the velocity (so that the speed was again `c`).

pdmullen replied:
Yeah happy to chat offline.  Early next week?

RyanWollaeger committed:
Apply suggested update for defrag task from Patrick.

Co-authored-by: Patrick Mullen <33095288+pdmullen@users.noreply.github.com>

RyanWollaeger committed:
Implement face sampling in new DDMC->IMC block-block function.

RyanWollaeger committed:
Mitigate apparent near-hangs from DDMC from ptcl_ddmc_albedo.

+ Reject particles eps_ddmc_offset back into IMC cells.
+ Check particle direction at cell boundary for entry condition.
+ Move particle eps_imc_offset*dt past end of time step if:
  + DDMC particle samples end time / census.
  + IMC particle is within eps_imc_offset*dt of end time / census.

RyanWollaeger commented:
I've run the CI script for darwin_volta-x86, but it won't seem to report (at least today).

RyanWollaeger committed:
Restrict volume source particles to an epsilon value inside cell.

RyanWollaeger commented:
Updated the MR title and description to summarize the later changes for the near-hangs.

Commit:
Stylistic changes

pdmullen commented on src/jaybenne/transport_utils.hpp:
```diff
@@ -147,6 +147,10 @@ void ptcl_transport_step(tran_step_args tra) {
   tra.y += tra.multi_d * tra.vy * dt_push;
   tra.z += tra.three_d * tra.vz * dt_push;
 
+  // push time slightly into next time step if particle is at end of time step (census)
+  const bool is_atend = fuzzy_equal(tra.t, tra.t_start + tra.dt, tra.dt, eps_imc_offset);
+  tra.t = is_atend ? tra.t_start + (1.0 + eps_imc_offset) * tra.dt : tra.t;
```
This change is also touching workflows invoking purely IMC (i.e., no DDMC).  For my own betterment, can you help me understand why it is required for pure IMC too?

RyanWollaeger replied:
It is a good question - this change is not strictly necessary for the main DDMC concerns of this MR. The reason I had put this in was out of a concern of hitting infinite iteration in the t+dt while-loop in the transport tasks. While not a major concern for doubles, if `Real`s were e.g. single-precision or lower, one could envision a scenario where the condition of the while-loop in the transport task is never met for a particle, since precision loss could cause the time to be lower than the t+dt repeatedly, when adding the particle time step to t.

pdmullen commented on src/jaybenne/transport_utils.hpp:
```diff
@@ -274,19 +279,42 @@ void ptcl_ddmc_step(ddmc_step_args dia) {
     dia.vx = dia.vv * nu * std::cos(phi);
     dia.vy = dia.vv * nu * std::sin(phi);
+
+    // ensure time is slightly past census time (TODO: use eps_ddmc_offset?)
+    dia.t = dia.t_start + (1.0 + eps_imc_offset) * dia.dt;
```
As above, also here?

RyanWollaeger replied:
The motivation here is the same as the motivation provided above. I can revert these changes since I did not observe these to impact the near-hang issue. Put a slightly different way, this is the time-analog of what we currently do in space, where we push the particle slightly into the spatial cell: here we are pushing it slightly into the next time step. With this change set, this should only be done (either for IMC or DDMC) when the minimal time/distance of the step (`dx_push`) is the time to the end of the time step.

pdmullen reviewed (approved):
LGTM! Thanks, @RyanWollaeger, for tracking this down.  I have a few questions above and committed a few stylistic changes.  Otherwise it seems ready to go!

Commit:
Fix my stylistic changes

RyanWollaeger committed:
Merge branch 'develop' of github.com:lanl/jaybenne into fix_crse_ddmc2imc
